### PR TITLE
Filtering feature

### PR DESCRIPTION
**Pull request by kalwalt**
This PR try to implement filtering of transpose matrix with [ArFilterTransMat](http://www.artoolkitx.org/artoolkit5/doc/apiref/arFilterTransMat_h/index.html#//apple_ref/c/func/arFilterTransMat) method. Probably with this code the interpolation routine that was implemented to smooth and remove the jittering of the model became un-necessary, i'm refering to this part of the code:
https://github.com/kalwalt/jsartoolkit5/blob/244b2b23286403e78fa24805b34509dc5a88052f/examples/nft_improved_worker/threejs_worker.js#L179-L191

with this PR simply became:

```js
 } else {
          sphere.visible = true; 
                // set matrix of 'root' by detected 'world' matrix
                setMatrix(root.matrix, world);
        }
```
**filterCutoffFrequency** i set to AR_FILTER_TRANS_MAT_CUTOFF_FREQ_DEFAULT that means has a default values of 5.0, in the nftSimple.c example instead is set to 15.0

**Build:**  build with emscripten Docker (latest)

### CODE REVIEW

File: emscripten/ARToolKitJS.cpp [hunk 1]
```diff
@@ -139,6 +144,14 @@ extern "C" {
 
 		if (arc->detectedPage >= 0) {
 			int trackResult = ar2TrackingMod(arc->ar2Handle, arc->surfaceSet[arc->detectedPage], arc->videoFrame, trans, &err);
+			for (int j = 0; j < 3; j++) {
+				for (int k = 0; k < 4; k++) {
+					transF[j][k] = trans[j][k];
+				}
+			}
+			if (arFilterTransMat(arc->ftmi, transF, 1) < 0) {
```
Comment (kalwalt) [inline emscripten/ARToolKitJS.cpp]: The third argument of  arFilterTransMat should a value that is zero or 1  as explained in the doc http://www.artoolkitx.org/artoolkit5/doc/apiref/arFilterTransMat_h/index.html#//apple_ref/c/func/arFilterTransMat
this needs to be changed.

**Comment by kalwalt** [comment]
i added a arLog to retrieve the value of the filterCutoffFrequency and it give me `0` so i think the code must be changed. So default value is `0`?

**Comment by kalwalt** [comment]
AR_FILTER_TRANS_MAT_SAMPLE_RATE_DEFAULT and AR_FILTER_TRANS_MAT_CUTOFF_FREQ_DEFAULT defined in arFilterTransMat.h:

https://github.com/artoolkitx/artoolkit5/blob/5bf0b671ff16ead527b9b892e6aeb1a2771f97be/include/AR/arFilterTransMat.h#L66-L81

as i wrote in the top comment. Maybe i was receiving that value because i haven't added them to ARBindEM.cpp

**Comment by nicolocarpignoli** [comment]
I will test it! @kalwalt does it work with threejs_gltf_worker example?

**Comment by kalwalt** [comment]
> I will test it! @kalwalt does it work with threejs_gltf_worker example?

works with all the nft examples but i removed the interpolation routine (that in the javscript code) in a dedicated example **main_threejs_filter_worker.html** try this and if you want the others. But i think i still need to fix few things to make some comparisons.

**Comment by kalwalt** [comment]
@nicolocarpignoli @ThorstenBux I added a WIP example with [gpu.js](https://github.com/gpujs/gpu.js) but i think that there are many variables in this approach.  I may implement a kernel for the matrix linear interpolation. Probably in this case that is not very recommended and i will remove it. Another idea is to implement a more efficient routine (linear interpolation) inside the C++ code. Probably will result faster?

**Comment by kalwalt** [comment]
@nicolocarpignoli @ThorstenBux i think the version with filtering and linear interpolation in the C++ code is much better, I'd like to know what you think :smile: 
Testing example: **examples/nft_improved_worker/main_threejs_filter_worker.html**

**Comment by kalwalt** [comment]
@nicolocarpignoli yes with the model animation, seems worste. I need to undersand why. Also fps are very slow on my Mobile device. Tested with a Wiko, View model with Android 7.1.2 and Chrome browser 79.0.3945.136

**Comment by kalwalt** [comment]
I will also remoove my gpu.js attempt, i think i will try to optimize the code without it, even if it sound interesting.

File: emscripten/ARToolKitJS.cpp [hunk 2]
```diff
@@ -99,6 +99,14 @@ extern "C" {
 		NFT API bindings
 	*/
 
+	void matrixLerp(ARdouble src[3][4], ARdouble dst[3][4], float interpolationFactor) {
```
Comment (kalwalt) [inline emscripten/ARToolKitJS.cpp:105]: We could try to use another lerp function, this maybe is not so efficient.
This function require **ARdouble**, not **float** data because **arFilterTransMat** function output **ARdouble**.
It also exist a lerp function from C++, we should see if we can use here in such a way.

File: emscripten/ARToolKitJS.cpp [hunk 3]
```diff
@@ -112,6 +120,8 @@ extern "C" {
 
 		float trans[3][4];
 		ARdouble transF[3][4];
+		ARdouble transFLerp[3][4];
+		memset( transFLerp, 0, 3 * 4 * sizeof(ARdouble) );
```
Comment (kalwalt) [inline emscripten/ARToolKitJS.cpp:137]: maybe this is not needed but not sure of this.

File: emscripten/ARToolKitJS.cpp [hunk 4]
```diff
@@ -160,6 +170,9 @@ extern "C" {
 			if (arFilterTransMat(arc->ftmi, transF, reset) < 0) {
 					ARLOGe("arFilterTransMat error with marker %d.\n", markerIndex);
 			}
+
+			matrixLerp(transF, transFLerp, 24.0);
```
Comment (kalwalt) [inline emscripten/ARToolKitJS.cpp]: try with another interpolationFactor, maybe 12.0 instead of 24.0?

File: emscripten/ARToolKitJS.cpp [hunk 5]
```diff
@@ -54,8 +54,8 @@ struct arController {
 	KpmHandle* kpmHandle;
 	AR2HandleT* ar2Handle;
 	ARFilterTransMatInfo *ftmi;
-	ARdouble   filterCutoffFrequency = AR_FILTER_TRANS_MAT_CUTOFF_FREQ_DEFAULT;
-	ARdouble   filterSampleRate = AR_FILTER_TRANS_MAT_SAMPLE_RATE_DEFAULT;
+	ARdouble   filterCutoffFrequency = 25.0;
```
Comment (kalwalt) [inline emscripten/ARToolKitJS.cpp]: try to increase filterCutoffFrequency

File: emscripten/ARToolKitJS.cpp [hunk 6]
```diff
@@ -54,8 +54,8 @@ struct arController {
...
+	ARdouble   filterSampleRate = 50.0;
```
Comment (kalwalt) [inline emscripten/ARToolKitJS.cpp]: and also filterSampleRate

File: emscripten/ARToolKitJS.cpp (hunk 2 above)
Comment (kalwalt) [reply emscripten/ARToolKitJS.cpp:105]: i am refering to this lerp https://en.cppreference.com/w/cpp/numeric/lerp

Comment (kalwalt) [reply emscripten/ARToolKitJS.cpp:105]: they say that interpolation factor should be a value between 0.0 and 1.0

Comment (kalwalt) [reply emscripten/ARToolKitJS.cpp:105]: see also this article on interpolation https://stackoverflow.com/questions/4353525/floating-point-linear-interpolation

File: emscripten/ARToolKitJS.cpp (hunk 4 above)
Comment (nicolocarpignoli) [reply emscripten/ARToolKitJS.cpp]: yes we have to try different values

File: emscripten/ARToolKitJS.cpp (hunk 2 above)
Comment (kalwalt) [reply emscripten/ARToolKitJS.cpp:105]: it seems that there are some incompatibility between std::lerp and emscripten, i can not use it. i will abandon this idea.

**Comment by kalwalt** [comment]
@nicolocarpignoli  i added another matrixLerp, if you want to test. It seems a bit better.

File: emscripten/ARToolKitJS.cpp [hunk 7]
```diff
@@ -274,6 +276,7 @@ extern "C" {
 		if (arControllers.find(id) == arControllers.end()) { return -1; }
 		arController *arc = &(arControllers[id]);
 		//arc->pixFormat = arVideoGetPixelFormat();
+		ARLOGi("filterCutoffFrequency value: %d\n", AR_FILTER_TRANS_MAT_CUTOFF_FREQ_DEFAULT);
```
Comment (kalwalt) [inline emscripten/ARToolKitJS.cpp]: This was wrong it should be: 

```diff
- ARLOGi("filterCutoffFrequency value: %d\n", AR_FILTER_TRANS_MAT_CUTOFF_FREQ_DEFAULT);
+ ARLOGi("filterCutoffFrequency value: %f\n", AR_FILTER_TRANS_MAT_CUTOFF_FREQ_DEFAULT);
```

File: emscripten/ARToolKitJS.cpp (hunk 3 above)
Comment (kalwalt) [reply emscripten/ARToolKitJS.cpp:137]: It is needed, without it the interpolation give weird results.

**Comment by kalwalt** [comment]
I think that would be optimal to let the user to build his own version with or without the filtering feature. I would do this feature adding some _#define_ to switch the filtering in the C++ code. So in the build process adding `-D WITH_FILTERING` it will be compiled with the filtering feature.

**Comment by kalwalt** [comment]
Testing example: https://kalwalt.github.io/kalwalt-interactivity-AR/nft/nft_threejs_filter_worker_gltf.html

**Comment by kalwalt** [comment]
I found in the BabylonAR repository this ['naive'](https://github.com/BabylonJS/BabylonAR/blob/master/src/cpp/aruco-meta-marker-tracker/src/NaiveFilter.h) implementation of the Kalman filter,  I should try it, it looks very promising!

**Comment by ThorstenBux** [comment]
Tested, 

duck is jittering like crazy

**Comment by evaristoc** [comment]
@kalwalt https://github.com/kalwalt/jsartoolkit5/pull/36#issuecomment-605684207 === this is a good way to go.

The Kalmar becomes a linear function after running for the first time so it can be even implemented in JS for some few relevant points.

Again, finding the right parameters might change by image and perhaps by device. But that I don't know.

If that the case, finding the right parameters might require a search algorithm (I have done some statistics / DS, so that my reference). But then you need to clarify which functions to optimise. I haven't done that job yet. A target success/failure variable is required.

The search can be automated if the right functions are formulated. The search might be slow and data intensive.

There are few more things that can be done to improve jittering that I have been testing. I will make a video and share the link to show you the results when having time. The target identification (short distance only) and homography were very good using the artoolkit algo for NFT. I will check later which algo is being used, I am really curious now. Previous essays I made with a simpler algos were not as good in detection, especially if the target image was a bit blurred, etc.

I will study some C++ shortly...

**Comment by evaristoc** [comment]
@kalwalt https://github.com/kalwalt/jsartoolkit5/pull/36#issuecomment-605684207

Nice reference! Thanks!

**Comment by evaristoc** [comment]
https://github.com/kalwalt/jsartoolkit5/pull/36#issuecomment-605693282

Yeah... the same here.

**Comment by kalwalt** [comment]
Hi @evaristoc recently i added a combined lerp and filtering (with arFilterTransMat) to my JsartoolkitNFT (dev branch) https://github.com/kalwalt/jsartoolkitNFT/pull/35 see if this is better or not. Also take a look at this discussion in the [artoolkitx/jsartoolkit5 issue tracker ](https://github.com/artoolkitx/jsartoolkit5/issues/97)

**Comment by evaristoc** [comment]
@kalwalt I check and let you know? Same testing link, I guess...

**Comment by evaristoc** [comment]
@kalwalt still jittering. Specially from some positions of the marker :disappointed:

**Comment by evaristoc** [comment]
@kalwalt  what is exactly arFilterTransMat? It is almost obvious from the name that it refers to the transformation matrix, but what does exactly do? Do you know?

EDIT: first order low pass filter: https://www.electronics-tutorials.ws/filter/filter_2.html:  "the low pass filter only allows low frequency signals from 0Hz to its cut-off frequency, ƒc point to pass while blocking those any higher."

It might be close to what it is done when practicing time series analysis in statistics. It looks like it is assuming a certain frequency (a sinusoidal signal) and implementing a linear model (first order) to clean it. It appears to work well for certain positions, likely because it can capture the signal.

**Comment by kalwalt** [comment]
> @kalwalt still jittering. Specially from some positions of the marker :disappointed:

Hi @evaristoc have you tested which code? 
**arFilterTransMat** is method coming from Artoolkit http://www.artoolkitx.org/artoolkit5/doc/apiref/arFilterTransMat_h/index.html (reference by ARtoolkitX but it is the same) read:
`High frequency noise ("jitter") can be an undesirable property in optical tracking systems like ARToolKit. These functions implement a first-order low-pass filter for both the position and orientation components of the pose estimate transformation matrices.`

Probably i didn't applyed correctly,It seems that has no effect ( or very low).

**Comment by evaristoc** [comment]
Hi @kalwalt I was editing my comment above:
EDIT: first order low pass filter: https://www.electronics-tutorials.ws/filter/filter_2.html: "the low pass filter only allows low frequency signals from 0Hz to its cut-off frequency, ƒc point to pass while blocking those any higher."

It might be close to what it is done when practicing time series analysis in statistics. It looks like it is assuming a certain frequency (a sinusoidal signal) and implementing a linear model (first order) to clean it. It appears to work well for certain positions, likely because it can capture the signal.

**Comment by evaristoc** [comment]
You might need not one but two or more of those to complete the job...

**Comment by kalwalt** [comment]
> Hi @kalwalt I was editing my comment above:
> EDIT: first order low pass filter: https://www.electronics-tutorials.ws/filter/filter_2.html: "the low pass filter only allows low frequency signals from 0Hz to its cut-off frequency, ƒc point to pass while blocking those any higher."
> 
> It might be close to what it is done when practicing time series analysis in statistics. It looks like it is assuming a certain frequency (a sinusoidal signal) and implementing a linear model (first order) to clean it. It appears to work well for certain positions, likely because it can capture the signal.

So maybe you can help, to find what is not correct in my approach. I Think it should filter the noise, but of course it should be done with a correct approach/method.

**Comment by evaristoc** [comment]
@kalwalt let's see. I expect to find a roof after which we can not improve the method any more, either because affects on performance or because it will mean to redo some aspects of the project.

Right now we are post-processing it.

But of course I want to help!

**Comment by evaristoc** [comment]
I would first advice to create a table or something visible of the inputs and outputs, @kalwalt?
You want to smooth the signal, so you might need several passes through different filters.

We should evaluate other filters too, as you have been mentioning in other repos.

**Comment by evaristoc** [comment]
@kalwalt will later make a sort of list of all the fronts where you are dealing with this (they are a few).

Let me know where you want to deal with this in particular. Here?

Meanwhile if you can point me to the script where you implemented this particular filter, the low pass one? We can probably create a factory for using not one but different kind of filters.

I suspect filters will have to be implemented case by case. I expect the jittering to change for different target images. I even expect that to change between devices, etc. But let's see, I need to research this more.

**Comment by kalwalt** [comment]
You can open an issue in this repository and fill all the data there and link to this PR. I prefere here because much control than in artoolkitx/jsartoolkit5. Eventually if yu are interested we have created a new org https://github.com/webarkit. If you are interested to join us, let me know! :slightly_smiling_face:

**Comment by kalwalt** [comment]
the **arFilterTransMat** is used in [ARToolKitJS,cpp](https://github.com/kalwalt/jsartoolkit5/blob/filtering-feature/emscripten/ARToolKitJS.cpp) the code is wrapped in `defines`:

```c++
#if WITH_FILTERING
// code here
#endif 
```
So you are not obliged to compile it, with filtering.

**Comment by evaristoc** [comment]
I will be having a look to different examples made with the same library. Example: http://jevois.org/

And keeping this one for further reading: https://answers.opencv.org/question/182075/convert-pose-from-artoolkit-into-opencv-screen-coordinates/

It seems to be some kind of overlap between the two. I have been doing some opencv recently too.

**Comment by evaristoc** [comment]
@kalwalt apparently this is [what ARfilterTransMat does](https://github.com/artoolkitx/artoolkit5/blob/5bf0b671ff16ead527b9b892e6aeb1a2771f97be/lib/SRC/AR/arFilterTransMat.c#L98)

The filter tries to correct orientation and position. For orientation it tries to keep the changes within a frequency. I think it is here where the actual low pass is done (**EDIT**: it is implemented in both, for orientation and for position, with the same calculated alpha, only that for a different equation (trig for orientation, linear for position)). For position is a simple lerp. It uses an alpha value for BOTH corrections and that for my perspective is a hmmm.....

That means you are probably implementing a second lerp over a lerp when going this route. (**EDIT**: I was referring here to the use of a custom LERP in other parts of the code you had been implementing.)

[alpha](https://github.com/artoolkitx/artoolkit5/blob/5bf0b671ff16ead527b9b892e6aeb1a2771f97be/lib/SRC/AR/arFilterTransMat.c#L93) is the corrector already explained in the material I passed to you.

It looks quite simple, IMHO.

I haven't checked at what point of the whole process flow it is implemented, but it is likely a post-processing, as we are all trying to do.

**Comment by evaristoc** [comment]
> You can open an issue in this repository and fill all the data there and link to this PR. I prefere here because much control than in artoolkitx/jsartoolkit5. Eventually if yu are interested we have created a new org https://github.com/webarkit. If you are interested to join us, let me know!

Yes, nice thanks!

**Comment by evaristoc** [comment]
@kalwalt 

Some additional observations about the ARfilterTransMat and jitter smoothing:

1. In the ARfilterTransMat the proposed sample rate is a fixed value of no particular use, IMO. I might be wrong, haven't seen all the implementation yet.
2. The resulting alpha value that will be used in the function as weighting is a constant. That is not necessarily a bad thing and it is the simplest implementation. However, variability of the noise signal might change from different positions, actual conditions (eg. lighting, device, position of the target image in front of the camera, etc)...
3. I have seen the ARfilterTransMat is being implemented in serious projects. That doesn't mean it is good though, but maybe convenient.
4. There is an interpolation used to correct positioning in the function. This is what the simple interpolation will do. Eg.

```
y0 = 10; //previous position
y1 = 20; //new position
alpha = 0.1; //I think it usually gives a low value in the function

y1Entry = alpha*y1 + (1-alpha)*y0 = 2 + 9 = 11 //interpolation
```

Similarly:

```
y0 = 10;
y1 = 0;
alpha = 0.1;

y1Entry = alpha*y1 + (1-alpha)*y0 = 0 + 9 = 9 //interpolation

```
In both cases, y1Entry will be used as y0 for the next frame.

5. In principle: you need to find a cut-off value that makes alpha small/big enough to smooth between y1 and y0. For what I have seen in NFT coordinates, the values of those coords are usually big, so you might have to try big numbers of cut-off value to see some effect.
6. Something about interpolation that might be useful: http://paulbourke.net/miscellaneous/interpolation/. A sort of interpolation could be a weighted moving average of two or more values, if I recall correctly.
7. The effect on the orientation smoothing by the ARfilterTransMat function is still to be checked.
8. I implemented a more discrete form of a pass filter [here](https://github.com/evaristoc/ar_experiments/blob/efebb2924997ffae646ea9e3d85b3b467108acff/0_apps/aframe-2dtracking/index.html#L743) together with [here](https://github.com/evaristoc/ar_experiments/blob/efebb2924997ffae646ea9e3d85b3b467108acff/0_apps/aframe-2dtracking/index.html#L767). _WARNING: Not without artifacts or performance costs!_. I think my implementation looks more like a band pass filter, not a low pass filter, and mine is sample-based. You can read it as follow:
> if new value changes over a previous value +/- cut-off, make new entry as _previous value_ +/- cut-off, otherwise make new entry equal new value.
In other words (I think...):
`y1Entry = min(y0 + cutoff, max(y0 - cutoff, y1))`

**Comment by kalwalt** [comment]
Thank you @evaristoc for the long deep comment, i will think about it. :smile:

**Comment by Carnaux** [comment]
@kalwalt how is the progress on this feature? Is it the same we talked about in the WebArkit slack?

**Comment by kalwalt** [comment]
@Carnaux haven't done some progress lately... Yes, this is in part what we talked in the WebARKit slack channel.
